### DISCUSSION TITLE
Linking issue: undefined reference to `xad::Tape<double>::Tape(bool)'

### DISCUSSION THREAD

**Discussion by bradbell** [auto-differentiation/XAD#120] [Q&A] [answered]
I think the the following bash script will produce the error message above (on any system that has g++).
Note that only the xad_d.a library gets installed (the xad.a library is not below the install prefix).

my_test.sh
```
#! /usr/bin/env bash
set -e -u
# -----------------------------------------------------------------------------
#
# xad.git
if [ ! -e xad.git ]
then
   git clone https://github.com/auto-differentiation/xad.git xad.git
fi
cd xad.git
#
# xad.git/build
if [ ! -e build ]
then
   mkdir build
fi
cd build
#
# xad.git/build/CMakeCache.txt
if [ -e CMakeCache.txt ]
then
   rm CMakeCache.txt
fi
#
# prefix
prefix=$(pwd)/prefix
#
#
# xad.git/build/cmake.out
# run cmake and put stdout and stderr in cmake.out
cmake -D CMAKE_INSTALL_PREFIX="$prefix" .. >& cmake.out
#
# xad.git/build/install.out
# run make install and stdout and stderr in install.out
make install >& install.out
#
# library_list
# find all the static libraries that have been installed
library_list=$(find $prefix -name '*.a')
echo "library_list = $library_list"
#
# xad.git/build/my_test.cpp
# creagte my_test.cpp, a simple program that uses xad
cat << EOF > my_test.cpp

# include <iostream>
# include <XAD/XAD.hpp>

template <class T>
T f(const T& x0)
{ return x0 * x0; }

int main()
{
    // input values
    double x0 = 2.0;

    // tape and active data type for 1st order adjoint computation
    typedef xad::adj<double> mode;
    typedef mode::tape_type tape_type;
    typedef mode::active_type AD;

    // initialize tape
    tape_type tape;

    // set independent variables
    AD x0_ad = x0;

    // and register them
    tape.registerInput(x0_ad);

    // start recording derivatives
    tape.newRecording();

    AD y = f(x0_ad);

    // register and seed adjoint of output
    tape.registerOutput(y);
    derivative(y) = 1.0;

    // compute all other adjoints
    tape.computeAdjoints();

    // output results
    std::cout << "y = " << value(y) << "\n"
              << "\nfirst order derivatives:\n"
              << "dy/dx0 = " << derivative(x0_ad) << "\n";
}
EOF
#
# xad.git/build/my_test.out
# attempt to compile and link the program above
if ! g++ \
   -g \
   -I $prefix/include \
   $library_list \
   -o my_test \
   my_test.cpp  >& my_test.out
then
   head my_test.out
   echo 'my_test.sh: Error'
   exit 1
fi
./my_test
echo 'my_test.sh: OK'
```

**Comment by auto-differentiation-dev** [comment]
Hi @bradbell, thanks for flagging this.

The XAD CMake build system is setup assuming that `CMAKE_BUILD_TYPE` is set at build time, to one of `Release`, `Debug`, `RelWithDebInfo`, or `MinSizeRel` (standard CMake configurations). We see in your script that none of these options is selected. The suffix `_d` in the library indicates that it assumed a Debug build, but it may not have been applied consistently throughout without setting the build type.

Could you try setting the build type and see if that fixes the issue? If not, you could check if the symbol is actually in the library? On Linux systems you can do that using 

```bash
nm -C path/to/libxad_d.a
```

And you should see if `xad::Tape<double>::Tape(bool)` is defined in it. (If yes, it is possible that the library file in your link line is not picked up for some reason and the linker therefore does not find the symbol)

Also note that you can build application libraries using CMake and link to XAD in an easier fashion. The CI/CD pipeline does that on every build on all platforms, as you can see [here](https://github.com/auto-differentiation/xad/blob/main/.github/workflows/ci.yml#L321-L330).

**Comment by bradbell** [comment, marked as answer]
@auto-differentiation-dev 
My mistake. I had my_test.cpp after the library in the compler/linker command; i.e.
```
 g++ \
   -g \
   -I $prefix/include \
   $library_list \
   -o my_test \
   my_test.cpp  >& my_test.out
   ```
   The following change fixed the problem:
   ```
   g++ \
   -g \
   -I $prefix/include \
   -o my_test \
   my_test.cpp  \
   $library_list \
   >& my_test.out
```
I suggest moving this from an issue to a discussion.
